pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -39,284 +108,413 @@ func Add(mgr manager.Manager) error {
+					return reconcile.Result{}, err
+				}
 
-	//Continue even if there are pods using emptyDir (local data that will be deleted when the node is drained).
-	//This is necessary for removing any pod that utilizes an emptyDir volume.
-	//The VirtualMachineInstance Pod does use emptryDir volumes,
-	//however the data in those volumes are ephemeral which means it is safe to delete after termination.
-	r.drainer.DeleteLocalData = true
+			}
 
-	//Ignore DaemonSet-managed pods.
-	//This is required because every node running a VirtualMachineInstance will also be running our helper DaemonSet called virt-handler.
-	//This flag indicates that it is safe to proceed with the eviction and to just ignore DaemonSets.
-	r.drainer.IgnoreAllDaemonSets = true
+		case LeaseStatusActiveOwnedByMeAboutToExpire:
+			{
+				if specAnnoData.enableMaintWindowExtensions {
 
-	//Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.
-	r.drainer.GracePeriodSeconds = -1
+					// renew the lease now.
+					lease.Spec.AcquireTime = &metav1.MicroTime{time.Now()}
 
-	// TODO - add logical value or attach from the maintancene CR
-	//The length of time to wait before giving up, zero means infinite
-	r.drainer.Timeout = time.Minute
+					if err := r.client.Update(context.TODO(), lease); err != nil {
+						r.reqLogger.Error("Failed to update the lease %v", err)
 
-	// TODO - consider pod selectors (only for VMIs + others ?)
-	//Label selector to filter pods on the node
-	//r.drainer.PodSelector = "kubevirt.io=virt-launcher"
+						statusAnnoData.status = NodeStateFailedToExtendLease
+						_, err = r.setStatusAnnotation(node, statusAnnoData)
 
-	cs, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return err
-	}
-	r.drainer.Client = cs
-	r.drainer.DryRun = false
+						return reconcile.Result{}, err
+					}
+				}
+			}
+			fallthrough
 
-	return nil
-}
+		case LeaseStatusActiveOwnedByMe:
+			{
 
-var _ reconcile.Reconciler = &ReconcileNodeMaintenance{}
+				if statusAnnoData.status != NodeStateCordon && statusAnnoData.status != NodeStateEvicting {
+					err := fmt.Errorf("Illegal state. node %s state %s", request.Name, statusAnnoData.status)
+					r.reqLogger.Errorf("error %v", err)
+					return reconcile.Result{}, err
+				}
 
-type ReconcileHandler interface {
-	StartPodInformer(node *corev1.Node, stop <-chan struct{}) error
-}
+				r.reqLogger.Infof("holding lease. Maintenance mode ongoing. node: %s\n", request.Name)
+			}
 
-// ReconcileNodeMaintenance reconciles a NodeMaintenance object
-type ReconcileNodeMaintenance struct {
-	// This client, initialized using mgr.Client() above, is a split client
-	// that reads objects from the cache and writes to the apiserver
-	client      client.Client
-	scheme      *runtime.Scheme
-	drainer     *drain.Helper
-	podInformer cache.SharedInformer
-}
+		case LeaseStatusActiveOwnedByDifferentNode:
```

Comment:
I don't understand, why would nodeA have a lease on nodeB?

This is a per-node lease, not a per-cluster one.
Owned by another controller (ie. machine healthcheck or MCO) is valid though

Reply:
renamed to LeaseStatusActiveOwnedByDifferentHolder  ; occurs if *lease.Spec.HolderIdentity of an existing lease is different from name of the current node.